### PR TITLE
Phase1 in FastSim of CMS

### PR DESCRIPTION
**Pull request by angirar**
New configurable tracker geometry for Phase0 and Phase1. Configurable track reconstruction sequence with new iterations added for Phase1. Iteration specific CA seeding (quadruplets and triplets) also in place.

### CODE REVIEW

**Review by makortel — COMMENTED** [review]
Thanks for the PR. It looks mostly good, but I found a couple of details more to be ironed out.

I believe it would be good to also add a 2017 FastSim workflow to `runTheMatrix.py` in this PR to allow testing of the developments.

File: RecoTracker/IterativeTracking/python/DetachedQuadStep_cff.py [hunk 1]
```diff
@@ -72,7 +73,20 @@
...
+    layerList = detachedQuadStepSeedLayers.layerList.value(),
+    BPix = cms.PSet(
+        TTRHBuilder = cms.string('WithoutRefit'),
+        HitProducer = cms.string('TrackingRecHitProducer'),
+        ),
+    FPix = cms.PSet(
+        TTRHBuilder = cms.string('WithoutRefit'),
+        HitProducer = cms.string('TrackingRecHitProducer'),
+        ),
+    layerPairs = detachedQuadStepHitDoublets.layerPairs.value(),
```
[the remark this answers is not in the document]
Comment (makortel) [reply RecoTracker/IterativeTracking/python/DetachedQuadStep_cff.py]: Could it be enough to use explicit types for `layerList` and `layerPairs`? I.e.
```python
layerList = cms.vstring(detachedQuadStepSeedLayers.layerList),
...
layerPairs = cms.vstring(detachedQuadStepHitDoublets.layerPairs),
```

[the remark this answers is not in the document]
Comment (angirar) [reply RecoTracker/IterativeTracking/python/DetachedQuadStep_cff.py]: I figured out the error. These parameters had to be added to the PSet of SeedFinderSelector here http://cmslxr.fnal.gov/source/FastSimulation/Tracking/python/TrajectorySeedProducer_cfi.py?v=CMSSW_10_2_X_2018-05-27-2300#0012
I have now removed the FastSim phase1 parameters from CA classes.

File: FastSimulation/Tracking/src/SeedFinderSelector.cc [hunk 2]
```diff
@@ -27,6 +26,10 @@ SeedFinderSelector::SeedFinderSelector(const edm::ParameterSet & cfg,edm::Consum
     , eventSetup_(nullptr)
     , measurementTracker_(nullptr)
     , measurementTrackerLabel_(cfg.getParameter<std::string>("measurementTracker"))
+    , layerList(cfg.getParameter<std::vector<std::string>>("layerList"))
+    , layerPairs(cfg.getParameter<std::vector<unsigned>>("layerPairs"))
+    , BPix(cfg.getParameter<edm::ParameterSet>("BPix"))
```
Comment (angirar) [inline FastSimulation/Tracking/src/SeedFinderSelector.cc]: The layer configuration for BPix and FPix could not be set by this method. I'm investigating it but it would be great to have some suggestions.

File: FastSimulation/Tracking/python/TrajectorySeedProducer_cfi.py [hunk 3]
```diff
@@ -10,7 +10,17 @@
     recHitCombinations = cms.InputTag("fastMatchedTrackerRecHitCombinations"),
     layerList = cms.vstring(),
     seedFinderSelector = cms.PSet(
-        measurementTracker = cms.string("")
+        measurementTracker = cms.string(""),
+        layerList = cms.vstring(""),
```
Comment (makortel) [inline FastSimulation/Tracking/python/TrajectorySeedProducer_cfi.py]: How is this `layerList` different from the `layerList` on line 11? From a very quick look on how they're used they seem to be set the same values. If they're indeed always the same, could one of the be dropped?

In addition, I think it would be more clear to remove the empty string arguments, i.e. to use `cms.string()` and `cms.vstring()`.

File: FastSimulation/Tracking/src/SeedFinderSelector.cc (hunk 2 above)
Comment (makortel) [reply FastSimulation/Tracking/src/SeedFinderSelector.cc]: I didn't quite understand what you mean, could you elaborate? The `BPix` member is an `edm::ParameterSet`, so there should be no problems is copy-constructing that here.

But why do you need the `layerList`, `layerPairs`, `BPix`, and `FPix` member variables? At the moment they are not used (and the `layerPairs` parameter is actually read to `layerPairs_` member variable here
https://github.com/cms-sw/cmssw/blob/230f31097d63e124ba340ecfc1662e55a15a65f3/FastSimulation/Tracking/src/SeedFinderSelector.cc#L46-L52

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by fabiocos** [comment]
please test

**Comment by slava77** [comment]
the errors are apparently related to this PR

```
    from RecoTracker.IterativeTracking.InitialStep_cff import *
  File "RecoTracker/IterativeTracking/InitialStep_cff.py", line 128, in <module>
    layerPairs = initialStepHitDoublets.layerPairs.value()
  File "FWCore/ParameterSet/Types.py", line 662, in clone
    _modifyParametersFromDict(myparams, params, self._Parameterizable__raiseBadSetAttr)
  File "FWCore/ParameterSet/Mixins.py", line 670, in _modifyParametersFromDict
    errorRaiser(key)
  File "FWCore/ParameterSet/Mixins.py", line 247, in __raiseBadSetAttr
    raise TypeError(name+" does not already exist, so it can only be set to a CMS python configuration type")
TypeError: layerPairs does not already exist, so it can only be set to a CMS python configuration type
```

I will wait for a working PR to review the reco changes

File: FastSimulation/Tracking/src/SeedFinderSelector.cc (hunk 2 above)
Comment (angirar) [reply FastSimulation/Tracking/src/SeedFinderSelector.cc]: Yes I do not need the new parameters as member variables. I have removed them.

File: FastSimulation/Tracking/python/TrajectorySeedProducer_cfi.py (hunk 3 above)
Comment (angirar) [reply FastSimulation/Tracking/python/TrajectorySeedProducer_cfi.py]: Actually the two layerList parameters are exactly the same. But the one in SeedFinderSelector PSet is required by SeedingLayerSetsBuilder,
https://github.com/angirar/cmssw/blob/1b9eea2212f6b52e5e9a91dad7a6cb16859a44d4/RecoTracker/TkSeedingLayers/src/SeedingLayerSetsBuilder.cc#L246
I will try to think if its possible to remove one of them.

File: RecoTracker/TkSeedingLayers/src/SeedingLayerSetsBuilder.cc [hunk 4]
```diff
@@ -181,11 +182,13 @@ std::string SeedingLayerSetsBuilder::LayerSpec::print(const std::vector<std::str
 SeedingLayerSetsBuilder::SeedingLayerSetsBuilder(const edm::ParameterSet & cfg, edm::ConsumesCollector&& iC):
   SeedingLayerSetsBuilder(cfg, iC)
 {}
-SeedingLayerSetsBuilder::SeedingLayerSetsBuilder(const edm::ParameterSet & cfg, edm::ConsumesCollector& iC)
+SeedingLayerSetsBuilder::SeedingLayerSetsBuilder(const edm::ParameterSet & cfg, edm::ConsumesCollector& iC):
+  isFastSim(cfg.getParameter<bool>("isFastSim"))
```
[the remark this answers is not in the document]
Comment (angirar) [reply RecoTracker/TkSeedingLayers/src/SeedingLayerSetsBuilder.cc]: Can you elaborate on this? Do you want me to create a new constructor which has the FastSim token as argument?

**Comment by angirar** [comment]
I have asked around for help on how to add a 2017 FastSim workflow to runTheMatrix.py file. Till then, here is the working recipe for generating FastSim phase1 sample (here TTBar) and running a trackingOnly validation - 
cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions 102X_mc2017_design_IdealBS_v1 --fast -n 5 --era Run2_2017_FastSim --datatier AODSIM,DQMIO --eventcontent AODSIM,DQM -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,RECO,VALIDATION:tracksValidationTrackingOnly --beamspot Realistic50ns13TeVCollision --python_filename fastsim.py --fileout fastsim.root

**Comment by slava77** [comment]
@angirar 
please post a message after you are done with updates of this PR based on already known comments or issues.
Thank you.

File: RecoTracker/TkSeedingLayers/src/SeedingLayerSetsBuilder.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (makortel) [reply RecoTracker/TkSeedingLayers/src/SeedingLayerSetsBuilder.cc]: Yes (although now I think maybe passing the collection name would be more clean and in line with what the `SeedingLayerSetsBuilder` does in standard reco). Note that you can call this constructor first to do it's job, so it should be something along
```cpp
SeedingLayerSetsBuilder::SeedingLayerSetsBuilder(const edm::ParameterSet & cfg, edm::ConsumesCollector& iC, const edm::InputTag& fastsimHitTag):
  SeedingLayerSetsBuilder(cfg, iC),
   fastSimrecHitsToken_(iC.consumes<FastTrackerRecHitCollection>(fastSimHitTag))
{}
```

File: FastSimulation/Tracking/python/TrajectorySeedProducer_cfi.py (hunk 3 above)
Comment (makortel) [reply FastSimulation/Tracking/python/TrajectorySeedProducer_cfi.py]: Could the higher-level `layerList` be removed and the piece which reads it be changed along
```diff
- iConfig.getParameter<std::vector<std::string>>("layerList")
+ iConfig.getParameter<edm::ParameterSet>("seedFinderSelector").getParameter<std::vector<std::string>>("layerList")
```
?

Please remove the "empty string" arguments from `cms.string` as well in here. And could `layerPairs` be just `cms.vuint32()`?

File: RecoPixelVertexing/PixelTriplets/src/CAHitTripletGenerator.cc [hunk 5]
```diff
@@ -57,7 +57,7 @@ void CAHitTripletGenerator::fillDescriptions(edm::ParameterSetDescription& desc)
   desc.add<double>("CAThetaCut", 0.00125);
   desc.add<double>("CAPhiCut", 0.1);
   desc.add<double>("CAHardPtCut", 0);
-
+
```
Comment (makortel) [inline RecoPixelVertexing/PixelTriplets/src/CAHitTripletGenerator.cc]: Could you undo this change, please?

File: RecoTracker/IterativeTracking/python/DetachedQuadStep_cff.py [hunk 6]
```diff
@@ -73,6 +74,7 @@
     CAThetaCut = 0.0011,
     CAPhiCut = 0,
 )
+
```
Comment (makortel) [inline RecoTracker/IterativeTracking/python/DetachedQuadStep_cff.py]: Could you undo this change, please?

File: RecoTracker/IterativeTracking/python/DetachedQuadStep_cff.py [hunk 7]
```diff
@@ -88,6 +90,23 @@
...
+    hitMasks = cms.InputTag("detachedQuadStepMasks"),
+    seedFinderSelector = dict( CAHitQuadrupletGeneratorFactory = _hitSetProducerToFactoryPSet(detachedQuadStepHitQuadruplets),
+                               #new parameters required for phase1 seeding
+                               layerList = detachedQuadStepSeedLayers.layerList.value(),
+                               BPix = dict(TTRHBuilder = 'WithoutRefit', HitProducer = 'TrackingRecHitProducer',),
+                               FPix = dict(TTRHBuilder = 'WithoutRefit', HitProducer = 'TrackingRecHitProducer',),
+                               layerPairs = detachedQuadStepHitDoublets.layerPairs.value()
+                               ))
+
+_fastSim_detachedQuadStepSeeds.seedFinderSelector.CAHitQuadrupletGeneratorFactory.SeedComparitorPSet.ComponentName = "none"
```
Comment (makortel) [inline RecoTracker/IterativeTracking/python/DetachedQuadStep_cff.py]: You could move this also in above with
```python
CAHitQuadrupletGeneratorFactory = _hitSetProducerToFactoryPSet(detachedQuadStepHitQuadruplets).clone(
   SeedComparitorPSet = dict(ComponentName = "none")
)
```

File: RecoTracker/IterativeTracking/python/DetachedTripletStep_cff.py [hunk 8]
```diff
@@ -114,8 +105,35 @@
     CAPhiCut = 0,
     CAHardPtCut = 0.2,
 ))
+
```
Comment (makortel) [inline RecoTracker/IterativeTracking/python/DetachedTripletStep_cff.py]: Could you undo this change, please?

File: RecoTracker/IterativeTracking/python/DetachedTripletStep_cff.py [hunk 9]
```diff
@@ -114,8 +105,35 @@
...
+_fastSim_detachedTripletStepSeeds = FastSimulation.Tracking.TrajectorySeedProducer_cfi.trajectorySeedProducer.clone(
+    layerList = detachedTripletStepSeedLayers.layerList.value(),
+    trackingRegions = "detachedTripletStepTrackingRegions",
+    hitMasks = cms.InputTag("detachedTripletStepMasks"),
+    seedFinderSelector = dict( pixelTripletGeneratorFactory = _hitSetProducerToFactoryPSet(detachedTripletStepHitTriplets))
+)
+#new for phase1
+trackingPhase1.toModify(_fastSim_detachedTripletStepSeeds, seedFinderSelector = dict(
+        pixelTripletGeneratorFactory = None,
+        CAHitTripletGeneratorFactory = _hitSetProducerToFactoryPSet(detachedTripletStepHitTriplets).clone(dict( SeedComparitorPSet = dict(ComponentName = "none"))),
```
Comment (makortel) [inline RecoTracker/IterativeTracking/python/DetachedTripletStep_cff.py]: I suspect the first `dict` is not strictly needed.

File: RecoTracker/IterativeTracking/python/InitialStep_cff.py [hunk 10]
```diff
@@ -71,14 +72,13 @@
     CAThetaCut = 0.0012,
     CAPhiCut = 0.2,
 )
+
```
Comment (makortel) [inline RecoTracker/IterativeTracking/python/InitialStep_cff.py]: Could you undo this change, please?

File: RecoTracker/IterativeTracking/python/InitialStep_cff.py [hunk 11]
```diff
@@ -71,14 +72,13 @@
...
 highBetaStar_2018.toModify(_initialStepCAHitQuadruplets,
     CAThetaCut = 0.0024,
     CAPhiCut = 0.4
 )
 initialStepHitQuadruplets = _initialStepCAHitQuadruplets.clone()
 
-trackingPhase1.toModify(initialStepHitDoublets, layerPairs = [0,1,2]) # layer pairs (0,1), (1,2), (2,3)
-
```
Comment (makortel) [inline RecoTracker/IterativeTracking/python/InitialStep_cff.py]: Could you undo this change, or explain why it is needed? (to me the new placement is not really better than this one)

File: RecoTracker/IterativeTracking/python/LowPtQuadStep_cff.py [hunk 12]
```diff
@@ -69,6 +70,7 @@
     CAThetaCut = 0.0017,
     CAPhiCut = 0.3,
 )
+
```
Comment (makortel) [inline RecoTracker/IterativeTracking/python/LowPtQuadStep_cff.py]: Could you undo this change, please?

File: RecoTracker/IterativeTracking/python/LowPtTripletStep_cff.py [hunk 13]
```diff
@@ -122,6 +112,7 @@
     CAPhiCut = 0.05,
 ))
 
+
```
Comment (makortel) [inline RecoTracker/IterativeTracking/python/LowPtTripletStep_cff.py]: Could you undo this change, please?

File: RecoTracker/IterativeTracking/python/LowPtTripletStep_cff.py [hunk 14]
```diff
@@ -137,7 +128,34 @@
...
+    hitMasks = cms.InputTag("lowPtTripletStepMasks"),
+)
 
+from FastSimulation.Tracking.SeedingMigration import _hitSetProducerToFactoryPSet
+_fastSim_lowPtTripletStepSeeds.seedFinderSelector.pixelTripletGeneratorFactory = _hitSetProducerToFactoryPSet(lowPtTripletStepHitTriplets)
+_fastSim_lowPtTripletStepSeeds.seedFinderSelector.pixelTripletGeneratorFactory.SeedComparitorPSet.ComponentName = "none"
+#new for phase1
+trackingPhase1.toModify(_fastSim_lowPtTripletStepSeeds, seedFinderSelector = dict(
+        pixelTripletGeneratorFactory = None,
+        CAHitTripletGeneratorFactory = _hitSetProducerToFactoryPSet(lowPtTripletStepHitTriplets).clone(dict( SeedComparitorPSet = dict(ComponentName = "none"))),
```
Comment (makortel) [inline RecoTracker/IterativeTracking/python/LowPtTripletStep_cff.py]: I suspect the first `dict` is not strictly needed.

File: RecoTracker/IterativeTracking/python/LowPtQuadStep_cff.py [hunk 15]
```diff
@@ -77,6 +79,23 @@
...
+    hitMasks = cms.InputTag("lowPtQuadStepMasks"),
+    seedFinderSelector = dict( CAHitQuadrupletGeneratorFactory = _hitSetProducerToFactoryPSet(lowPtQuadStepHitQuadruplets),
+                               #new parameters required for phase1 seeding
+                               layerList = lowPtQuadStepSeedLayers.layerList.value(),
+                               BPix = dict(TTRHBuilder = 'WithoutRefit', HitProducer = 'TrackingRecHitProducer',),
+                               FPix = dict(TTRHBuilder = 'WithoutRefit', HitProducer = 'TrackingRecHitProducer',),
+                               layerPairs = lowPtQuadStepHitDoublets.layerPairs.value()
+                               ))
+
+_fastSim_lowPtQuadStepSeeds.seedFinderSelector.CAHitQuadrupletGeneratorFactory.SeedComparitorPSet.ComponentName = "none"
```
Comment (makortel) [inline RecoTracker/IterativeTracking/python/LowPtQuadStep_cff.py]: You could move this also in above with
```python
CAHitQuadrupletGeneratorFactory = _hitSetProducerToFactoryPSet(lowPtQuadStepHitQuadruplets).clone(
   SeedComparitorPSet = dict(ComponentName = "none")
)
```

File: RecoTracker/TkSeedingLayers/src/SeedingLayerSetsBuilder.cc [hunk 16]
```diff
@@ -185,7 +186,7 @@ SeedingLayerSetsBuilder::SeedingLayerSetsBuilder(const edm::ParameterSet & cfg,
 {
   std::vector<std::string> namesPset = cfg.getParameter<std::vector<std::string> >("layerList");
   std::vector<std::vector<std::string> > layerNamesInSets = this->layerNamesInSets(namesPset);
-
+  fastSimrecHitsToken_ = iC.consumes<FastTrackerRecHitCollection>(edm::InputTag("fastTrackerRecHits")); //fix: to pass as argument to constructor
```
Comment (makortel) [inline RecoTracker/TkSeedingLayers/src/SeedingLayerSetsBuilder.cc]: (This is probably on your list already, but to not forget) please move this to a FastSim-specific constructor which takes the `edm::InputTag` as an argument as discussed in https://github.com/cms-sw/cmssw/pull/23363#discussion_r192752766

**Review by makortel — COMMENTED** [review]
Thanks. I have now only mostly cosmetic comments.

(I guess the FastSim-specific constructor for `SeedingLayerSetsBuilder` is already on your list to do)

File: RecoTracker/TkSeedingLayers/src/SeedingLayerSetsBuilder.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (angirar) [reply RecoTracker/TkSeedingLayers/src/SeedingLayerSetsBuilder.cc]: I tried it this way but I keep getting the following error - 
Exception Message:
A get using a EDGetToken with the C++ type 'edm::OwnVector<FastTrackerRecHit,edm::ClonePolicy<FastTrackerRecHit> >' was made using an uninitialized token.
 Please check that the variable is being initialized from a 'consumes' call.

I'm investigating it.

[the remark this answers is not in the document]
Comment (makortel) [reply RecoTracker/TkSeedingLayers/src/SeedingLayerSetsBuilder.cc]: Do you call that constructor from the FastSim side?

File: Configuration/PyReleaseValidation/python/relval_steps.py [hunk 17]
```diff
@@ -875,6 +875,18 @@ def genS(fragment,howMuch):
 step1FastUpg2015_trackingOnlyValidation = merge([{'-s':'GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,RECO,VALIDATION:@trackingOnlyValidation'},
                                                 step1FastUpg2015Defaults])
 
+step1FastUpg2017Defaults =merge([{'-s':'GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION:@standardValidation,DQM:@standardDQM',
+                           '--fast':'',
+                           '--conditions'  :'auto:phase1_2017_realistic',
+                           '--beamspot'    : 'Realistic50ns13TeVCollision',
```
Comment (fabozzi) [inline Configuration/PyReleaseValidation/python/relval_steps.py]: @angirar : note that in Phase 1 MC the beasmpot is Realistic25ns13TeVEarly2017Collision (you can check the fullsim relvals)

Comment (angirar) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: Made the change. Thanks.

File: RecoTracker/TkSeedingLayers/src/SeedingLayerSetsBuilder.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (angirar) [reply RecoTracker/TkSeedingLayers/src/SeedingLayerSetsBuilder.cc]: Yes I am. Infact, when I declared the constructor the way you have written above, it gives error : "mem-initializer for 'SeedingLayerSetsBuilder::fastSimrecHitsToken_' follows constructor delegation
   fastSimrecHitsToken_(iC.consumes<FastTrackerRecHitCollection>(fastsimHitTag))" 
So instead I declared it simply as 
SeedingLayerSetsBuilder::SeedingLayerSetsBuilder(const edm::ParameterSet & cfg, edm::ConsumesCollector& iC, const edm::InputTag& fastsimHitTag):
  fastSimrecHitsToken_(iC.consumes<FastTrackerRecHitCollection>(fastsimHitTag))
{} 
which works, and I'm calling it from SeedFinderSelector like this - "SeedingLayerSetsBuilder s(cfg, consumesCollector, edm::InputTag("fastTrackerRecHits"));" after making the unique pointer of SeedingLayerSetsBuilder here https://github.com/angirar/cmssw/blob/9fb1e01e75e58cef6411dfae4d642918442b3d72/FastSimulation/Tracking/src/SeedFinderSelector.cc#L54
I know it goes inside the new constructor and comes back (by putting the cout statements) but still gives an error about 'uninitialized token'.

[the remark this answers is not in the document]
Comment (angirar) [reply RecoTracker/TkSeedingLayers/src/SeedingLayerSetsBuilder.cc]: Ah right.. with little tweaking its working now. Thanks.

**Comment by makortel** [comment]
@cmsbuild, please test workflow 2017.1

Let's give it a try.

File: FastSimulation/Tracking/python/TrajectorySeedProducer_cfi.py (hunk 3 above)
Comment (angirar) [reply FastSimulation/Tracking/python/TrajectorySeedProducer_cfi.py]: Yes, all done except for when I try to remove the "empty strings" arguments from cms.string I get the following error -
Exception Message:
python encountered the error: <type 'exceptions.TypeError'>
__init__() takes exactly 2 arguments (1 given)

Comment (makortel) [reply FastSimulation/Tracking/python/TrajectorySeedProducer_cfi.py]: That sounds like a problem elsewhere. Could you run `python <your-CMSSW-config>` instead of `cmsRun`? That will give the full stack trace of the failure.

File: RecoTracker/IterativeTracking/python/DetachedTripletStep_cff.py [hunk 18]
```diff
@@ -116,6 +107,31 @@
...
+_fastSim_detachedTripletStepSeeds = FastSimulation.Tracking.TrajectorySeedProducer_cfi.trajectorySeedProducer.clone(
+    trackingRegions = "detachedTripletStepTrackingRegions",
+    hitMasks = cms.InputTag("detachedTripletStepMasks"),
+    seedFinderSelector = dict( pixelTripletGeneratorFactory = _hitSetProducerToFactoryPSet(detachedTripletStepHitTriplets),
+                               layerList = detachedTripletStepSeedLayers.layerList.value())
+)
+#new for phase1
+trackingPhase1.toModify(_fastSim_detachedTripletStepSeeds, seedFinderSelector = dict(
+        pixelTripletGeneratorFactory = None,
+        CAHitTripletGeneratorFactory = _hitSetProducerToFactoryPSet(detachedTripletStepHitTriplets).clone(SeedComparitorPSet = dict(ComponentName = "none")),
```
Comment (makortel) [inline RecoTracker/IterativeTracking/python/DetachedTripletStep_cff.py]: Looking the test failures, `detachedTripletStepHitTriplets` does not have `SeedComparitorPSet` parameter, so you should drop the customizations inside the `clone()` call.

Comment (angirar) [reply RecoTracker/IterativeTracking/python/DetachedTripletStep_cff.py]: Right. I will remove it.

File: FastSimulation/Tracking/python/TrajectorySeedProducer_cfi.py (hunk 3 above)
Comment (angirar) [reply FastSimulation/Tracking/python/TrajectorySeedProducer_cfi.py]: This is the output of 'python <config.py>' :
Traceback (most recent call last):
  File "fastsim.py", line 24, in <module>
    process.load('FastSimulation.Configuration.Reconstruction_BefMix_cff')
  File "/cvmfs/cms.cern.ch/slc6_amd64_gcc630/cms/cmssw/CMSSW_10_2_0_pre3/python/FWCore/ParameterSet/Config.py", line 582, in load
    module = __import__(moduleName)
  File "/afs/cern.ch/work/a/arastogi/private/PhaseIUpgrade_Studies/25May2018/CMSSW_10_2_0_pre3/python/FastSimulation/Configuration/Reconstruction_BefMix_cff.py", line 42, in <module>
    from FastSimulation.Tracking.iterativeTk_cff import *
  File "/afs/cern.ch/work/a/arastogi/private/PhaseIUpgrade_Studies/25May2018/CMSSW_10_2_0_pre3/python/FastSimulation/Tracking/iterativeTk_cff.py", line 9, in <module>
    from RecoTracker.IterativeTracking.InitialStep_cff import *
  File "/afs/cern.ch/work/a/arastogi/private/PhaseIUpgrade_Studies/25May2018/CMSSW_10_2_0_pre3/python/RecoTracker/IterativeTracking/InitialStep_cff.py", line 109, in <module>
    import FastSimulation.Tracking.TrajectorySeedProducer_cfi
  File "/afs/cern.ch/work/a/arastogi/private/PhaseIUpgrade_Studies/25May2018/CMSSW_10_2_0_pre3/python/FastSimulation/Tracking/TrajectorySeedProducer_cfi.py", line 13, in <module>
    measurementTracker = cms.string(),
TypeError: __init__() takes exactly 2 arguments (1 given)

Comment (makortel) [reply FastSimulation/Tracking/python/TrajectorySeedProducer_cfi.py]: Right, on a closer look `cms.string()` indeed requires an argument, so the empty strings have to be kept. Sorry for the noise...

**Comment by makortel** [comment]
@angirar Have you tested locally that a FastSim workflow works now?

**Comment by angirar** [comment]
This was the error -
----- Begin Fatal Exception 06-Jun-2018 15:06:53 CEST-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'validation_step'
   [2] Calling method for module SiPixelPhase1TrackClustersV/'SiPixelPhase1TrackClustersAnalyzerV'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: edmNew::DetSetVector<SiPixelCluster>
Looking for module label: siPixelClusters
Looking for productInstanceName: 

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------

**Comment by makortel** [comment]
And FastSim doesn't produce clusters, right? Then just disable it (with a comment explaining why).

**Comment by angirar** [comment]
@makortel Okay thanks. I did it and step1 ran successfully. However, there exists segmentation fault in the step2 i.e. in the validationHarvesting step. I'm looking at it now for the source of error.

**Comment by pmaksim1** [comment]
@makortel, all: I can definitively confirm that FastSim does not produce SiPixelClusters.  It goes from PSimHits to SiPixelRecHits directly.

**Comment by civanch** [comment]
please test

**Comment by makortel** [comment]
@cmsbuild, please test workflow 2017.1

Let's test also the added code.

**Comment by kpedro88** [comment]
In step1 of the new workflow, there are a lot of messages like this:
```
%MSG-e LowPtClusterShapeSeedComparitor:   TrajectorySeedProducer:initialStepSeeds 11-Jun-2018 10:04:30 CEST  Run: 1 Event: 10
this is not a pixel cluster
%MSG
```
is this expected?

**Comment by pmaksim1** [comment]
@kpedro88 
The relevant code from 
http://cmslxr.fnal.gov/source/RecoPixelVertexing/PixelLowPtUtilities/src/LowPtClusterShapeSeedComparitor.cc#0153
is:

0152     const SiPixelRecHit* pixelRecHit =
0153       dynamic_cast<const SiPixelRecHit *>(hits[i]->hit());
0154 
0155     if (!pixelRecHit){
0156       edm::LogError("LowPtClusterShapeSeedComparitor")<<"this is not a pixel cluster";
0157       ok = false; break;
0158     }

Since FastSim uses real SiPixelRecHits, the dynamic_cast<> should work.  So something is not right.

Now having said this, I don't think we should be running this in FastSim at all!  The idea behind the LowPtClusterShapeSeedComparitor is that it uses cluster shapes to filter out the seeds, since the cluster shape is strongly correlated with the local track angles, and thus one can throw away seeds where the local angles of the seed are incompatible with the cluster shapes.

The problem here is that, in FastSim, we actually don't have clusters, and thus the cluster shape information is unavailable.  We *could* do that (since the part of FastSim which simulates SiPixelRecHits actually produces an average cluster shape, but then throws it away), but we are not doing it at the moment.

Thus, right now, even if we manage this code to work, it's actually not doing anything useful.

**Comment by pmaksim1** [comment]
This was with my FastSim hat on.  Wearing the offline pixel hat, I should also add that the way this code is setup is very crude, and will fail with radiation damage -- there is a better way of doing that.  But it's a discussion for a RECO meeting in (not very near) future.

File: RecoTracker/IterativeTracking/python/InitialStep_cff.py [hunk 19]
```diff
@@ -106,11 +106,28 @@
...
     trackingRegions = "initialStepTrackingRegions",
-    seedFinderSelector = dict( pixelTripletGeneratorFactory = _hitSetProducerToFactoryPSet(initialStepHitTriplets))
+    seedFinderSelector = dict( pixelTripletGeneratorFactory = _hitSetProducerToFactoryPSet(initialStepHitTriplets),
+                               layerList = initialStepSeedLayers.layerList.value())
 )
 _fastSim_initialStepSeeds.seedFinderSelector.pixelTripletGeneratorFactory.SeedComparitorPSet.ComponentName = "none"
+#new for phase1
+trackingPhase1.toModify(_fastSim_initialStepSeeds, seedFinderSelector = dict(
+        pixelTripletGeneratorFactory = None,
+        CAHitQuadrupletGeneratorFactory = _hitSetProducerToFactoryPSet(initialStepHitQuadruplets).clone(dict( SeedComparitorPSet = dict(ComponentName = "none"))),
```
Comment (makortel) [inline RecoTracker/IterativeTracking/python/InitialStep_cff.py]: @angirar You should remove the first `dict`, i.e. do `clone(SeedComparitorPSet = dict(...`.

(thanks @pmaksim1 for the analysis)

I'll see if framework could be made to throw an exception in this kind of cases as passing `dict` as an argument to `clone()` doesn't make much sense(?).

**Comment by makortel** [comment]
@cmsbuild, please test workflow 2017.1

**Comment by makortel** [comment]
Now the `LowPtClusterShapeSeedComparitor` printouts are gone. There is, however, one related to HCAL
```
%MSG-e HcalDigitizer:  MixingModule:mix 12-Jun-2018 11:21:52 CEST  Run: 1 Event: 3
bad hcal id found in digitizer. Skipping 1159218239 (HE 17,63,1)
%MSG
```
(only one occurrence in 10 events)

**Comment by kpedro88** [comment]
@makortel this has to do with the merging of depths for the HEP17 test installation from 2017. The cell `(HE 17,63,1)` is a "merged" one, so it should not come out of the SIM step... we might have to wait for @bsunanda to check this out. But it should be addressed at some point.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/23363	b88c1b8
- changes appear to be in line with the PR description and (from what I lightly followed) with the review by @makortel . There are no changes in the standard full pp reco or in the phase-0 fastsim.
- jenkins tests pass and comparisons with the baseline show no relevant differences
    - fastsim tests 135.4 (2016) and 5.1 (Run1) show some changes in seedingLayerSet plots, but these appear to be technical (due to reordering of seeding layer set driven by the configuration).

**Comment by makortel** [comment]
> changes appear to be in line with the PR description and (from what I lightly followed) with the review by @makortel

Right, maybe I could note explicitly that with b88c1b8 I'm fine with this PR.

**Comment by kpedro88** [comment]
+1
followup item: check how fastsim is handling merged DetIds for HCAL HEP17

**Comment by dmitrijus** [comment]
+1

**Comment by ssekmen** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by fabiocos** [comment]
+operations

the addition of a new Era and the corresponding modification to the standard sequences appear consistent with the purpose of the PR

**Comment by fabiocos** [comment]
@fabozzi @prebello could you please check and sign it in case?

**Comment by fabozzi** [comment]
+1

**Comment by fabiocos** [comment]
+1
